courseworks/week8/TianhaoHU/hw8.ipynb

```diff
@@ -0,0 +1,185 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 68,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "using Optim"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 69,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "create_loss_function (generic function with 1 method)"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "function create_loss_function(vertex_count, edges, locations)\n",
+    "    function loss(x)\n",
+    "        total_loss = 0.0\n",
+    "        for i in 1:vertex_count\n",
+    "            for j in (i+1):vertex_count\n",
+    "                xi, xj = x[2*i-1:2*i], x[2*j-1:2*j]\n",
+    "                dist = sqrt(sum((xi - xj).^2))\n",
+    "                connected = any([(i, j) == edge || (j, i) == edge for edge in edges])\n",
+    "\n",
+    "                if connected && dist > 1.0\n",
+    "                    total_loss += (dist - 1.0)^2\n",
+    "                elseif !connected && dist < 1.0\n",
+    "                    total_loss += (1.0 - dist)^2\n",
+    "                end\n",
```

Comment:
here you directly defined $ r = 1 $ as the condition, however for the unitdisk embeding I will recommand you that you may choos 0.95 as requirement for 'connected' and verse vice.